A47-xds-federation.md

```diff
@@ -0,0 +1,693 @@
+- It allows multiple authorities to use the same server channel, via the
+  de-duplication described above.
+- It will pave the way for future work where each authority can specify
+  a list of servers, ordered by priority.  (In other words, the
+  authority entry in the cache will change which server channel it is
+  using based on which server is currently reachable.)
+
+The second-level map will be the resource type, exactly the same as the
+top-level map is today.
+
+The third-level map will be the resource name.  For old-style resource
+names, this will be the full resource name, exactly as it is today.  For
+new-style names, it will be the `id` part of the `xdstp:` URI (i.e., the
+URI path with the resource type prefix stripped off) and the context
+parameters.
+
+For example, a resource with old-style name `server.example.com` and
+type Listener might be stored like this:
+
+```
+{authority "old:" =>
+  {resource_type Listener =>
+    {"server.example.com" => <resource>}
+  }
+}
+```
+
+But a resource with new-style name
+`xdstp://xds.authority.com/envoy.config.listener.v3.Listener/server.example.com?foo=bar`
+might be stored like this:
+
+```
+{authority "xdstp:xds.authority.com" =>
+  {resource_type Listener =>
+    {("server.example.com", {"foo"=>"bar"}) => <resource>}
+  }
+}
+```
+
+New-style resource names used as cache keys must be
+[normalized](https://github.com/cncf/xds/blob/main/proposals/TP1-xds-transport-next.md#normalization).
+In particular, this means sorting the context params in normal lexicographic
+order.
+
+If a new-style name has duplicate values for the same context param, the
+behavior is undefined.  The easiest thing is probably to just use the
+last value seen, since this allows the implementation to just populate
+the map by iterating over the query params.
+
+##### Watcher Changes
+
+When a caller starts a watch on an old-style resource name (one that
+does not start with `xdstp:`), the `XdsClient` will behave essentially as
+it does today, using the empty string as the authority name in the
+resource cache.
+
+When a caller starts a watch on a new-style resource name (one that
+starts with `xdstp:`), the `XdsClient` will parse the URI.  It will use
+the authority of the URI as the primary key in the resource cache (using
+the empty string if the URI has no authority).  It will canonify any
+context parameters that are part of the resource name (i.e., removing
+duplicate keys and sorting keys in lexicographic order).  It will then
+check the cache to see if the resource is already present, in which case
+it will be returned immediately.  Otherwise, it will send a message
+subscribing to that resource, using the appropriate xDS server from the
+authority.
+
+#### LRS Server Representation
+
+We currently represent the LRS server name in several places:
+- In the `CdsUpdate` struct returned by the `XdsClient` CDS watcher.
+- In the LB policy configs for the `xds_cluster_resolver` and
```

Comment:
The field (same in both balancers), is called `lrsLoadReportingServerName`.
Do we keep using that field? Or rename it? Or add a new field?

Reply:
Good question.  This made me realize that we should really spell out the exact LB policy config changes, and that made me notice a few problems.  The latest commit has a concrete proposal.

Reply:
I've sent grpc/grpc-proto#101 with the service config changes, as well as a couple of other missing updates.

Reply:
CC @dapengzhang0 as FYI for Java impl.